sys/usb/usbus/cdc/acm/cdc_acm_stdio.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2018 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup sys
+ * @{
+ *
+ * @file
+ * @brief CDC ACM stdio implementation for USBUS CDC ACM
+ *
+ * This file implements a USB CDC ACM callback and read/write functions.
+ *
+ *
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "isrpipe.h"
+
+#include "usb/usbus.h"
+#include "usb/usbus/cdc/acm.h"
+
+#if MODULE_VFS
+#include "vfs.h"
+#endif
+
+static usbus_cdcacm_device_t cdcacm;
+static uint8_t _cdc_tx_buf_mem[USBUS_CDC_ACM_STDIO_BUF_SIZE];
+static uint8_t _cdc_rx_buf_mem[USBUS_CDC_ACM_STDIO_BUF_SIZE];
+isrpipe_t cdc_stdio_isrpipe = ISRPIPE_INIT(_cdc_rx_buf_mem);
+
+void stdio_init(void)
+{
+    /* Initialize this side of the CDC ACM pipe */
+#if MODULE_VFS
+    vfs_bind_stdio();
+#endif
+}
+
+ssize_t stdio_read(void* buffer, size_t len)
+{
+    (void)buffer;
+    (void)len;
+    return isrpipe_read(&cdc_stdio_isrpipe, buffer, len);
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    usbus_cdc_acm_submit(&cdcacm, buffer, len);
+    usbus_cdc_acm_flush(&cdcacm);
+    /* Use tsrb and flush */
+    return len;
+}
+
+static void _cdc_acm_rx_pipe(usbus_cdcacm_device_t *cdcacm,
+                             uint8_t *data, size_t len)
+{
+    (void)cdcacm;
+    for (size_t i = 0; i < len; i++) {
+        isrpipe_write_one(&cdc_stdio_isrpipe, data[i]);
+    }
+}
+
+void usb_cdc_acm_stdio_init(usbus_t *usbus)
+{
+    usbus_cdc_acm_init(usbus, &cdcacm, _cdc_acm_rx_pipe, NULL,
```

Comment:
How could the coding_cb function could be updated (it's NULL by default) after the CDC ACM stdio is initialized ? I'm thinking of the Arduino MKR (and any board flashed via bossa) where we could provide a way to automatically reset the board to bootloader mode when the host open the TTY with a given baudrate: This is what is done on Arduino: when the stdio serial is opened at 1200 baud, the board reboot in bootloader mode and bossa can flash the board after that.
Maybe add a setter function to the API for this ? Maybe I'm missing something obvious in the STDIO initialization workflow.

Reply:
> How could the coding_cb function could be updated (it's NULL by default) after the CDC ACM stdio is initialized ?

There is no update function provided at the moment, I can add it if it is required.

The aim of this callback is to allow the host computer to configure the serial device, setting the baud rate and the parity mode. For STDIO this normally doesn't make much sense as there is no baud rate or parity associated with normal STDIO operations. With #12268 this is used to configure and update the baud rate and mode of the `periph_uart` device used.

> I'm thinking of the Arduino MKR (and any board flashed via bossa) where we could provide a way to automatically reset the board to bootloader mode when the host open the TTY with a given baudrate: This is what is done on Arduino: when the stdio serial is opened at 1200 baud, the board reboot in bootloader mode and bossa can flash the board after that.

So the idea here is to reset (`pm_reboot()`) the board when the ttyACMn connected to the RIOT USB serial console is configured to a specific baud rate? I guess I can add an update hook function if you want to have this structure as a way to reset the board.

Reply:
>  when the ttyACMn connected to the RIOT USB serial console is configured to a specific baud rate? I guess I can add an update hook function

That would be useful indeed.

Reply:
> That would be useful indeed.

Added something, please let me know if this is what you had in mind.

Reply:
Yes, exactly. Not sure if it's the best solution though ;)

Reply:
It should be working (I didn't test it), maybe we can leave it as is now and optimize later?